File: templates/problems/11_graphs/08_course_schedule/python/solution.py
def solve_can_finish(num_courses: int, prerequisites: list[list[int]]) -> bool:
    """Detect if all courses can be finished (no cycle) using DFS.

    Time: O(V + E), Space: O(V + E).
    """
    graph: list[list[int]] = [[] for _ in range(num_courses)]
    for p in prerequisites:
        graph[p[1]].append(p[0])

    # 0=unvisited, 1=visiting, 2=visited
    state = [0] * num_courses

    def has_cycle(node: int) -> bool:
        if state[node] == 1:
            return True
        if state[node] == 2:
            return False
        state[node] = 1
        for nxt in graph[node]:
            if has_cycle(nxt):
                return True
        state[node] = 2
        return False

    for i in range(num_courses):
        if has_cycle(i):
            return False
    return True

File: templates/problems/11_graphs/08_course_schedule/typescript/solution.ts
/**
 * Detect if all courses can be finished (no cycle) using DFS.
 * Time: O(V + E), Space: O(V + E).
 */
export function solveCanFinish(numCourses: number, prerequisites: number[][]): boolean {
    const graph: number[][] = Array.from({ length: numCourses }, () => []);
    for (const p of prerequisites) {
        graph[p[1]].push(p[0]);
    }

    // 0=unvisited, 1=visiting, 2=visited
    const state = new Array(numCourses).fill(0);

    const hasCycle = (node: number): boolean => {
        if (state[node] === 1) return true;
        if (state[node] === 2) return false;
        state[node] = 1;
        for (const nxt of graph[node]) {
            if (hasCycle(nxt)) return true;
        }
        state[node] = 2;
        return false;
    };

    for (let i = 0; i < numCourses; i++) {
        if (hasCycle(i)) return false;
    }
    return true;
}
